### DISCUSSION TITLE
Qt的rules使用场景覆盖不全

### DISCUSSION THREAD

**Discussion by ThzShun** [xmake-io/xmake#4114] [Q&A]
### 你在什么场景下需要该功能？

现有设计，没法解决如下以下需求：
1. 库 仅用到QtCore和moc
2. 库 使用了QtWidgets，但是不是一个可执行程序（即不是 kind="binary"）
3. rules没法传导到依赖的上层target中。
如A库是一个用qt widget 的静态库，B是依赖A的一个可执行目标。B需要`add_rules("qt.widgetapp")`才可以。
4. add_frameworks 也同样无法传到上层target中。

此外还有个问题，和Qt相关的就不开多一个issue了。
5. 在有`add_rules("qt.moc")`的target中，`add_files("src/**.h")`是否可以支持递归处理，并且没有`Q_OBJECT`的不进行moc文件生成。

### 描述可能的解决方案

```
我可能预想的改进设计：
- 项目依赖的qt项，还是走add_deps，但是多一个参数，来处理复杂的include、link等
- 自由组合 kind的 static、shared、binary 和 widget/qml的关系
```

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
> 库 仅用到QtCore和moc
库 使用了QtWidgets，但是不是一个可执行程序（即不是 kind="binary"）

走 qt.shared ，额外配置 QtWidgets frameworks 不就好了，没啥本质区别。。

```lua
target("xxx")
    add_rules("qt.shared")
    add_frameworks("QtWidgets")
```

qt.widgetapp 也就是比它多了一些内置 frameworks 而已，你可以自己配。。qt.static 也是

> rules没法传导到依赖的上层target中。
> 如A库是一个用qt widget 的静态库，B是依赖A的一个可执行目标。B需要add_rules("qt.widgetapp")才可以。

自己 B 里面配置下 `add_rules("qt.widgetapp")` ，这个没啥好纠结的，目前设计如此，改不了。。

> add_frameworks 也同样无法传到上层target中。

设置下 `{public = true}` 就就行了。。`add_frameworks("xxx", {public = true})`

> 5. 在有add_rules("qt.moc")的target中，add_files("src/**.h")是否可以支持递归处理，并且没有Q_OBJECT的不进行moc文件生成。

现在不就是这样的么。。还需要改啥？ 没有 Q_OBJECT 的干嘛要加 add_files 里。。自己去掉。
